pkg/rbac/rbac.go

```diff
@@ -0,0 +1,314 @@
+//
+// - kClient is K8s cluster client. This should be set when API consumer intents to use
+// the AccessReviewer instance for a single user. The provided k8s client connection will be directly
+// used to fetch ACLs from the cluster. In this case, access review  API can be invoked without needing
+// to pass the user's Token on every call.
+func NewAccessReviewer(kConfig *rest.Config, kClient kubernetes.Interface) (*AccessReviewer, error) {
+	// Verify only one of k8s config or client are set
+	if kClient == nil && kConfig == nil {
+		return nil, errors.New("one of either kubeConfig or kubeClient must be a non-nil value")
+	}
+
+	if kClient != nil && kConfig != nil {
+		return nil, errors.New("only one  of either kubeConfig or kubeClient must be a non-nil value")
+	}
+
+	accessReviewer := new(AccessReviewer)
+
+	if kConfig != nil {
+		configCopy := *kConfig
+		accessReviewer.kubeConfig = &configCopy
+	} else {
+		accessReviewer.kubeClient = kClient
+	}
+
+	return accessReviewer, nil
+}
+
+// getKubeClientForUser returns the k8s client to use to connect to the cluster.
+// - userToken is the user's OAuth bearer token.  It will be used along with the k8sConfig,
+// set on the AccessReviewer, to create a new k8s client. If k8sConfig is not available,
+// then the configured k8s client is returned.
+func (r *AccessReviewer) getKubeClientForUser(userToken string) (kubernetes.Interface, error) {
+	if r.kubeConfig != nil {
+		// if a valid userToken
+		if userToken != "" {
+			// make a copy of the RestConfig to avoid overwrites when multiple api calls are made in parallel
+			userKubeConfig := rest.CopyConfig(r.kubeConfig)
+			userKubeConfig.BearerToken = userToken
+			// create the clientset
+			kclient, err := kubernetes.NewForConfig(userKubeConfig)
+			if err != nil {
+				return nil, err
+			}
+
+			return kclient, nil
+		}
+
+		return nil, fmt.Errorf(
+			"failed to get a client to connect to the kubernetes cluster:" +
+				"When KubeConfig is provided, a valid userToken must be set on all access review calls")
+	}
+
+	// if kubeConfig isnt set then return the kubeClient set
+	return r.kubeClient, nil
+}
+
+// GetMetricsAccess retrieves the user's ACLs from the k8s cluster  and processes them to determine
+// user's access to observability metrics that are gathered from managed clusters.
+// It returns a map of allowed managedclusters and the allowed namespaces for each managedcluster.
+//
+// - userToken is the user's OAuth bearer token, is required if k8s config was set on the AccessReviewer
+//
+// - clusters are the  names of the managedclusters for which  allowed metrics access is processed and returned.
+// if no clusters are specified, then  metrics access is returned for all "allowed" managedclusters.
+func (r *AccessReviewer) GetMetricsAccess(userToken string, clusters ...string) (map[string][]string, error) {
+	klog.V(2).Infof("GetMetricsAccess for clusters: %v", clusters)
+
+	// get Client to talk to the Kubernetes cluster
```

Comment:
Optional: In general, comments that are pointing out the obvious aren't a best practice since they can get stale over time and become inaccurate.